core/src/credential.rs

```diff
@@ -0,0 +1,147 @@
+
+type CredentialSubject = serde_json::Value;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct VerifiableCredential <'a> {
+    #[serde(rename = "@context")]
+	context:  VCContext,
+    #[serde(rename = "@id")]
+	id: String,
+    cred_type:  String,
+	issuer: String,
+	issuance_date: SystemTime,
+	subject: CredentialSubject,
+}
+
+pub struct IssuedCredential <'a> {
+    verifiable_credential:VerifiableCredential,
+    proof: *mut CredentialProof,
+}
+
+pub struct CredentialManager {
+    credential_subject: HashMap<&str, &str>,
+    context: VCContext,
+}
+
+trait VC {
+    fn getSubject(&self, key: &str) -> Result<String, Error> {
+        match self.credential_subject.get(key) {
+            Some(&value) => Ok(value.to_string()),
+            None => Err(Error::new(ErrorKind::Other, "Key not found")),
+        } 
+    }
+    fn getContext(&self) -> VCContext {
+        self.context.clone()
+    }
+}
+
+/// Default VC trait implementation for CredentialManager.
+/// Callers may re-implement this trait in their code by: impl VC for CredentialManager {/...code here.../}
+impl VC for CredentialManager {}
+
+impl <'a> CredentialManager <'a> {
+    /// returns CredentialManager cred_subject_template with default configuration accessible for overlapping via VC trait
+    pub fn new() -> CredentialManager {
+        ///default credential subjects for different credential types 
+        let cred_subject_template: HashMap<&str, &str> = HashMap::from([
+                (CRED_TYPE_PERMANENT_RESIDENT_CARD, r###"{
+                    "@id": "https://w3id.org/citizenship#PermanentResidentCard",
+                    "@context": {
+                      "@version": 1.1,
+                      "@protected": true,
+                      "id": "@id",
+                      "type": "@type",
+                      "description": "http://schema.org/description",
+                      "name": "http://schema.org/name",
+                      "identifier": "http://schema.org/identifier",
+                      "image": {"@id": "http://schema.org/image", "@type": "@id"}
+                    }
+                  }"###),
+                (CRED_TYPE_BANK_CARD, r###"{
+                    "account":"",
+                    "address":"",
+                    "birthDate":"",
+                    "branch":"",
+                    "country":"",
+                    "familyName":"",
+                    "gender":"",
+                    "givenName":"",
+                    "id":"",
+                    "phone":"",
+                    "type":[
+                       "BankCard"
+                    ]
+                 }"###),
+            ]);
+        return Self { 
+            credential_subject: cred_subject_template,
+            context: JSON_LD_CONTEXT,
+         }
+    }
+
+    pub fn initVerifiableCredential (&self, cred_type: String, cred_subject: serde_json::Value, issuer: &'a str, id: &'a str) -> Result(VerifiableCredential, Error) {
```

Comment:
This function should receive the subject and context via parameters. The `DocumentBuilder` will use the override functions. I think our actual credential creation functionality should be as "pure" and stateless as possible in order to make it easier to test and debug

Reply:
So we separate the "data collection" i.e. defining what pieces we'll use to create the VC. And the actual process of creation itself.

Gathering -> Construction